### PR TITLE
Check system storage rebate conservation at epoch boundary

### PR DESCRIPTION
**Pull request by lxfind**
At epoch boundary, if a node config flag is enabled, we iterate through all live objects, sum their storage rebate, and compare it to the storage fund balance, make sure they matches.
In order to ship this to testnet, however, we will need to know (after we fix all the conservation issues) the balance gap and compensate that manually.

### CODE REVIEW

File: crates/sui-node/src/lib.rs [hunk 1]
```diff
@@ -1115,6 +1119,44 @@ impl SuiNode {
         assert_eq!(next_epoch, new_epoch_store.epoch());
         new_epoch_store
     }
+
+    fn check_system_consistency(&self) {
+        if !self.config.enable_expensive_safety_checks && cfg!(not(debug_assertions)) {
```
Comment (sblackshear) [inline crates/sui-node/src/lib.rs:1124]: Will this run on the `stress` + `simtest` nightly (which I think runs in debug mode and might get rather expensive)? Not necessarily suggesting we should turn it off, but worth understanding.

File: crates/sui-node/src/lib.rs [hunk 2]
```diff
@@ -1115,6 +1119,44 @@ impl SuiNode {
...
+            return;
+        }
+        let total_storage_rebate: u64 = self
+            .state
+            .db()
+            .iter_live_object_set()
+            .map(|o| {
+                if let LiveObject::Normal(object) = o {
+                    object.storage_rebate
```
Comment (sblackshear) [inline crates/sui-node/src/lib.rs:1133]: Should we go ahead and go `object.get_total_sui()` here? That we can also check that the total is 10B and we'll cover everything.

File: crates/sui-node/src/lib.rs [hunk 3]
```diff
@@ -1015,6 +1016,9 @@ impl SuiNode {
                     None
                 }
             } else {
+                // TODO: Enable this when we can.
```
Comment (sblackshear) [inline crates/sui-node/src/lib.rs:1019]: Looks a bit confusing to have the TODO here when the gating happens at 1124.

File: crates/sui-node/src/lib.rs (hunk 2 above)
Comment (lxfind) [reply crates/sui-node/src/lib.rs:1133]: Yes we can. Although that will really depend on everything to be fixed first

File: crates/sui-node/src/lib.rs [hunk 4]
```diff
@@ -1115,6 +1119,44 @@ impl SuiNode {
...
+    fn check_system_consistency(&self) {
+        if !self.config.enable_expensive_safety_checks && cfg!(not(debug_assertions)) {
+            return;
+        }
+        let total_storage_rebate: u64 = self
+            .state
+            .db()
+            .iter_live_object_set()
+            .map(|o| {
+                if let LiveObject::Normal(object) = o {
```
Comment (sblackshear) [inline crates/sui-node/src/lib.rs:1132]: Exhaustive pattern matching might save us some grief if we have multiple ways to represent a `LiveObject` w/ a storage rebate in the future.

**Review by sblackshear — APPROVED** [review]

File: crates/sui-node/src/lib.rs (hunk 1 above)
Comment (lxfind) [reply crates/sui-node/src/lib.rs:1124]: Yes. It will, assuming debug_assertions is enabled in that

**Review by williampsmith — APPROVED** [review]
LGTM on the state accumulator / `LiveObject` front

**Pull request merged**